### PR TITLE
chore(E2E) : dashboard creation while editing dashboard

### PR DESCRIPTION
**Pull request by MedWassCh**
## Description

dashboard creation while editing dashboard

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x (master)

### CODE REVIEW

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts [hunk 1]
```diff
@@ -18,6 +17,7 @@ after(() => {
   cy.stopWebContainer();
 });
 
+
```
Comment (kduret) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:20]:
```suggestion
```

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts [hunk 2]
```diff
@@ -195,3 +195,38 @@ Then('the form fields are empty', () => {
...
+});
+
+Then("creates a new dashboard on the previous dashboard's edition page", () => {
+      cy.getByTestId({ testId: 'ArrowDropDownIcon' }).click();
+      cy.getByTestId({ testId: 'AddIcon' }).click();
+      cy.getByLabel({ label: 'Name', tag: 'input' }).type(dashboards.default.name);
+      cy.getByLabel({ label: 'Description', tag: 'textarea' }).type(dashboards.default.description);
+      cy.getByLabel({ label: 'Create', tag: 'button' }).click()
+      cy.wait('@createDashboard')
+
```
Comment (kduret) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:219]: linting seems wrong : indentation, carriage return..

File: centreon/packages/js-config/cypress/e2e/commands.ts [hunk 3]
```diff
@@ -399,6 +399,38 @@ Cypress.Commands.add(
   }
 );
 
+Cypress.Commands.add('insertDashboardWithPatch', (dashboardBody, patchBody) => {
```
Comment (paul-oureib) [inline centreon/packages/js-config/cypress/e2e/commands.ts:402]: the function itself is correct and good, but you probably should find a more appropriate name for the function (like insertDashboardWithWidget)

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts [hunk 4]
```diff
@@ -195,3 +195,38 @@ Then('the form fields are empty', () => {
...
+});
+
+Then("the user is redirected to the newly created dashboard's edition page", () => {
+  cy.url().should((url) => {
+    const urlStr = url.toString();
+    expect(urlStr).to.include('/dashboard');
+    expect(urlStr).to.include('edit=true');
+  });
```
Comment (kduret) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:227]:
```suggestion
  cy.url().should('match', /\/dashboard.+edit=true/);
```

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts [hunk 5]
```diff
@@ -195,3 +195,38 @@ Then('the form fields are empty', () => {
   cy.getByLabel({ label: 'Name', tag: 'input' }).should('be.empty');
   cy.getByLabel({ label: 'Description', tag: 'textarea' }).should('be.empty');
 });
+
+Given(
+  "a dashboard with existing widgets in the dashboard administrator user's dashboard library",
+  () => {
+    cy.insertDashboardWithPatch(dashboards.fromDashboardCreatorUser,patchBody);
```
Comment (kduret) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:202]:
```suggestion
    cy.insertDashboardWithPatch(dashboards.fromDashboardCreatorUser ,patchBody);
```

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts [hunk 6]
```diff
@@ -1,8 +1,7 @@
 import { Given, Then, When } from '@badeball/cypress-cucumber-preprocessor';
 import { last } from 'ramda';
-
 import dashboards from '../../../fixtures/dashboards/creation/dashboards.json';
-
+import patchBody from '../../../fixtures/dashboards/creation/widgets/patchBody.json'
```
Comment (kduret) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:4]: keep blank line between external libraries import and local import

File: centreon/packages/js-config/cypress/e2e/commands.ts (hunk 3 above)
Comment (kduret) [inline centreon/packages/js-config/cypress/e2e/commands.ts:402]: `insertDashboardWithPatch` is a weird name
Choose a more appropriate one like `insertDashboardDuringEditionOfExistingOne`

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts [hunk 7]
```diff
@@ -195,3 +195,38 @@ Then('the form fields are empty', () => {
...
+  "a dashboard with existing widgets in the dashboard administrator user's dashboard library",
+  () => {
+    cy.insertDashboardWithPatch(dashboards.fromDashboardCreatorUser,patchBody);
+    cy.visit(`${Cypress.config().baseUrl}/centreon/home/dashboards`);
+  }
+);
+
+
+When('the dashboard administrator user starts to edit the dashboard', () => {
+      cy.getByLabel({ label: 'view', tag: 'button' }).click();
```
Comment (kduret) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:209]:
```suggestion
  cy.getByLabel({ label: 'view', tag: 'button' }).click();
```

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts [hunk 8]
```diff
@@ -195,3 +195,38 @@ Then('the form fields are empty', () => {
...
+Then("the user is redirected to the newly created dashboard's edition page", () => {
+  cy.url().should((url) => {
+    const urlStr = url.toString();
+    expect(urlStr).to.include('/dashboard');
+    expect(urlStr).to.include('edit=true');
+  });
+});
+
+Then("the newly created dashboard is empty", () => {
+  cy.getByTestId({ testId: 'AddIcon' }).should('be.visible')
```
Comment (kduret) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:231]: I am not sure the assertion does what is corresponding to the step edition
I guess there should be something like `should('not.be.visible')` or `should('not.exist')`

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts (hunk 2 above)
Comment (paul-oureib) [reply centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:219]: you should probably install eslint on your IDE

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts (hunk 8 above)
Comment (paul-oureib) [reply centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:231]: i think the object referred to with the AddIcon testId only appears whenever there'sq no widget on the dashboard, Wassim should confirm this but it probably has its place there

Comment (kduret) [reply centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:231]: There are mutiple AddIcon in the page, so this assertion is not enough : 
image

File: centreon/packages/js-config/cypress/e2e/commands.ts [hunk 9]
```diff
@@ -485,6 +520,11 @@ declare global {
       hoverRootMenuItem: (rootItemNumber: number) => Cypress.Chainable;
       insertDashboard: (dashboard: Dashboard) => Cypress.Chainable;
       insertDashboardList: (fixtureFile: string) => Cypress.Chainable;
+      insertDashboardWithWidget: (
+        dashboard: Dashboard,
+        patch: any
```
Comment (kduret) [inline centreon/packages/js-config/cypress/e2e/commands.ts:525]: can you define a stricter type (declare an interface)

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts [hunk 10]
```diff
@@ -1,8 +1,9 @@
 import { Given, Then, When } from '@badeball/cypress-cucumber-preprocessor';
 import { last } from 'ramda';
-
```
Comment (kduret) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation/index.ts:3]: restore carriage return

File: centreon/packages/js-config/cypress/e2e/commands.ts [hunk 11]
```diff
@@ -413,6 +448,30 @@ interface ListingRequestResult {
   };
 }
 
+interface PatchBody {
```
Comment (kduret) [inline centreon/packages/js-config/cypress/e2e/commands.ts:451]: PatchBody is too generic
```suggestion
interface PatchDashboardBody {
```

File: centreon/packages/js-config/cypress/e2e/commands.ts [hunk 12]
```diff
@@ -485,6 +544,11 @@ declare global {
       hoverRootMenuItem: (rootItemNumber: number) => Cypress.Chainable;
       insertDashboard: (dashboard: Dashboard) => Cypress.Chainable;
       insertDashboardList: (fixtureFile: string) => Cypress.Chainable;
+      insertDashboardWithWidget: (
+        dashboard: Dashboard,
+        patch: PatchBody
```
Comment (kduret) [inline centreon/packages/js-config/cypress/e2e/commands.ts:549]:
```suggestion
        patch: PatchDashboardBody
```

**Review by kduret — DISMISSED** [review]

**Review by paul-oureib — DISMISSED** [review]

**Review by kduret — APPROVED** [review]

**Review by sc979 — DISMISSED** [review]
LGTM as all @paul-oureib comments seem taken care of

**Review by Thebarda — APPROVED** [review]

File: centreon/tests/e2e/features/Dashboards/01-dashboard-creation.feature [hunk 13]
```diff
@@ -31,4 +31,12 @@ Feature: Creating a new dashboard
     Then the user is on the dashboards overview page
     And the dashboard has not been created
     When the user opens the form to create a new dashboard for the second time
-    Then the form fields are empty
\ No newline at end of file
+    Then the form fields are empty
+
+  @TEST_MON-23634
+  Scenario: Creating a new dashboard while editing a dashboard
```
Comment (bdauria) [inline centreon/tests/e2e/features/Dashboards/01-dashboard-creation.feature:37]:
```suggestion
  Scenario: Creating a new dashboard while editing an existing one
```

**Review by bdauria — APPROVED** [review]

**Review by sc979 — APPROVED** [review]

**Review by paul-oureib — APPROVED** [review]

**Review by arnapou — APPROVED** [review]

**Pull request merged**